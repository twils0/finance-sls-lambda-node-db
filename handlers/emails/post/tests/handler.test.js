const { loadPool } = require('../../../../functions/db/loadPool');
const { testAuth } = require('../../../../functions/aws/testAuth');
const { updateUser } = require('../../../../functions/db/updateUser');
const { sendTemplatedEmail } = require('../../../../functions/aws/ses/sendTemplatedEmail');

const { post } = require('../handler');

global.console.log = jest.fn();
jest.mock('../../../../functions/db/loadPool', () => ({
  loadPool: jest.fn(),
}));
jest.mock('../../../../functions/aws/testAuth', () => ({
  testAuth: jest.fn(),
}));
jest.mock('../../../../functions/db/updateUser', () => ({
  updateUser: jest.fn(),
}));
jest.mock('../../../../functions/aws/ses/sendTemplatedEmail', () => ({
  sendTemplatedEmail: jest.fn(),
}));
jest.mock('../../../../functions/errorResponse', () => ({
  errorResponse: jest.fn(error => error),
}));

const accessToken = { test: 'testAccessToken' };
const userId = 'testUserID';
const template = 'verificationEmailTemplate';
const email = 'test1@test.com';
const emailAdditional = 'test2@test.com';
const emailVerificationId = '43663f5d-a27c-4e1f-8d22-0b0f42b1b27a';
const emailAdditionalVerificationId = '7cfcfcc9-0e50-4a0b-90a7-d340a2e51960';
const emailNeededEvent = {
  query: {
    accessToken,
  },
  cognitoPoolClaims: {
    user_id: userId,
    email,
    email_ver: 'false',
    email_additional: null,
    email_additional_ver: 'true',
  },
};
const emailAdditionalNeededEvent = {
  query: {
    accessToken,
  },
  cognitoPoolClaims: {
    user_id: userId,
    email,
    email_ver: 'true',
    email_additional: emailAdditional,
    email_additional_ver: 'false',
  },
};
const emailEmailAdditionalNeededEvent = {
  query: {
    accessToken,
  },
  cognitoPoolClaims: {
    user_id: userId,
    email,
    email_ver: 'false',
    email_additional: emailAdditional,
    email_additional_ver: 'false',
  },
};

global.pool = { test: 'testPool' };
const response = {
  status: 200,
  body: {
    message: 'success',
  },
};

const callback = jest.fn();

describe('handlers', () => {
  describe('emails', () => {
    describe('post', () => {
      afterEach(() => {
        loadPool.mockReset();
        testAuth.mockReset();
        updateUser.mockReset();
        sendTemplatedEmail.mockReset();
        global.console.log.mockReset();
        callback.mockReset();
      });

      it('fails and returns error, when loadPool throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.reject(error));

        await post(emailNeededEvent, null, callback);

        expect(callback).toBeCalledWith(error, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('fails and returns error, when testAuth throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.reject(error));

        await post(emailNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(callback).toBeCalledWith(error, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('fails and returns error, when updateUser throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.resolve());
        updateUser.mockReturnValue(Promise.reject(error));

        await post(emailNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(updateUser).toBeCalledWith(userId, {
          emailVerified: 'false',
        });
        expect(sendTemplatedEmail).not.toBeCalled();
        expect(callback).toBeCalledWith(error, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('fails and returns error, when sendTemplatedEmail throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.resolve());
        updateUser.mockReturnValue(Promise.resolve({ emailVerificationId }));
        sendTemplatedEmail.mockReturnValue(Promise.reject(error));

        await post(emailNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(updateUser).toBeCalledWith(userId, {
          emailVerified: 'false',
        });
        expect(sendTemplatedEmail).toBeCalledWith(
          template,
          { verificationId: emailVerificationId },
          [email],
        );
        expect(callback).toBeCalledWith(error, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('correctly sends templated email or emails, with email needed', async () => {
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.resolve());
        updateUser.mockReturnValue(Promise.resolve({ emailVerificationId }));

        await post(emailNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(updateUser).toBeCalledWith(userId, {
          emailVerified: 'false',
        });
        expect(sendTemplatedEmail).toBeCalledWith(
          template,
          { verificationId: emailVerificationId },
          [email],
        );
        expect(callback).toBeCalledWith(null, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('correctly sends templated email or emails, with email additional needed', async () => {
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.resolve());
        updateUser.mockReturnValue(Promise.resolve({ emailAdditionalVerificationId }));

        await post(emailAdditionalNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(updateUser).toBeCalledWith(userId, {
          emailVerified: 'true',
          emailAdditionalVerified: 'false',
        });
        expect(sendTemplatedEmail).toBeCalledWith(
          template,
          { verificationId: emailAdditionalVerificationId },
          [emailAdditional],
        );
        expect(callback).toBeCalledWith(null, response);
        expect(global.console.log).toMatchSnapshot();
      });

      it('correctly sends templated email or emails, with email and email additional needed', async () => {
        loadPool.mockReturnValue(Promise.resolve());
        testAuth.mockReturnValue(Promise.resolve());
        updateUser.mockReturnValue(Promise.resolve({
          emailVerificationId,
          emailAdditionalVerificationId,
        }));

        await post(emailEmailAdditionalNeededEvent, null, callback);

        expect(testAuth).toBeCalledWith(accessToken);
        expect(updateUser).toBeCalledWith(userId, {
          emailVerified: 'false',
          emailAdditionalVerified: 'false',
        });
        expect(sendTemplatedEmail).toBeCalledWith(
          template,
          { verificationId: emailAdditionalVerificationId },
          [emailAdditional],
        );
        expect(sendTemplatedEmail).toBeCalledWith(
          template,
          { verificationId: emailVerificationId },
          [email],
        );
        expect(callback).toBeCalledWith(null, response);
        expect(global.console.log).toMatchSnapshot();
      });
    });
  });
});
