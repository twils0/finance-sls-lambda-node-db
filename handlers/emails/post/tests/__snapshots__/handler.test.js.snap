// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handlers emails post correctly sends templated email or emails, with email additional needed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "updateUser
",
      Object {
        "emailAdditionalVerified": "false",
        "emailVerified": "true",
      },
      "
",
    ],
    Array [
      "updateUser - success
",
      undefined,
      "
",
      "7cfcfcc9-0e50-4a0b-90a7-d340a2e51960",
      "
",
    ],
    Array [
      "sendTemplatedEmail
",
      "verificationEmailTemplate",
      "
",
      Object {
        "verificationId": "7cfcfcc9-0e50-4a0b-90a7-d340a2e51960",
      },
      "
",
      Array [
        "test2@test.com",
      ],
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post correctly sends templated email or emails, with email and email additional needed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "updateUser
",
      Object {
        "emailAdditionalVerified": "false",
        "emailVerified": "false",
      },
      "
",
    ],
    Array [
      "updateUser - success
",
      "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      "
",
      "7cfcfcc9-0e50-4a0b-90a7-d340a2e51960",
      "
",
    ],
    Array [
      "sendTemplatedEmail
",
      "verificationEmailTemplate",
      "
",
      Object {
        "verificationId": "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      },
      "
",
      Array [
        "test1@test.com",
      ],
      "
",
    ],
    Array [
      "sendTemplatedEmail
",
      "verificationEmailTemplate",
      "
",
      Object {
        "verificationId": "7cfcfcc9-0e50-4a0b-90a7-d340a2e51960",
      },
      "
",
      Array [
        "test2@test.com",
      ],
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post correctly sends templated email or emails, with email needed 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "updateUser
",
      Object {
        "emailVerified": "false",
      },
      "
",
    ],
    Array [
      "updateUser - success
",
      "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      "
",
      undefined,
      "
",
    ],
    Array [
      "sendTemplatedEmail
",
      "verificationEmailTemplate",
      "
",
      Object {
        "verificationId": "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      },
      "
",
      Array [
        "test1@test.com",
      ],
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post fails and returns error, when loadPool throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post fails and returns error, when sendTemplatedEmail throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "updateUser
",
      Object {
        "emailVerified": "false",
      },
      "
",
    ],
    Array [
      "updateUser - success
",
      "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      "
",
      undefined,
      "
",
    ],
    Array [
      "sendTemplatedEmail
",
      "verificationEmailTemplate",
      "
",
      Object {
        "verificationId": "43663f5d-a27c-4e1f-8d22-0b0f42b1b27a",
      },
      "
",
      Array [
        "test1@test.com",
      ],
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post fails and returns error, when testAuth throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`handlers emails post fails and returns error, when updateUser throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAuth
",
    ],
    Array [
      "loadPool
",
    ],
    Array [
      "updateUser
",
      Object {
        "emailVerified": "false",
      },
      "
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
